### PR TITLE
[functorch] Fix grad/vjp/jvp returning zeros under inference_mode

### LINKED ISSUE #177318: torch.func transforms silently return zeros under inference_mode
## 🐛 Describe the bug

  `torch.func.{grad, vjp, jvp}` silently produce incorrect results (zeros)
  when called under `torch.inference_mode()`. No error is raised.

  ```python
  import torch
  from torch.func import jvp

  x = torch.randn(3)
  with torch.inference_mode():
      out, tangent = jvp(lambda x: x**2, (x,), (torch.ones(3),))
      print(tangent)  # tensor([0., 0., 0.]) — should be 2*x
```
##  Root cause

  Functorch transforms set up their own autograd context on entry:
  enable_grad() for grad/vjp, _set_fwd_grad_enabled(True) for jvp
  (see NOTE [grad and vjp interaction with no_grad]). However,
  inference_mode also excludes autograd dispatch keys from the TLS
  dispatch key set, and the mode-enabling calls don't undo that. So
  autograd dispatch keys remain excluded and differentiation silently
  produces zeros.

  The transforms already correctly handle no_grad — this is the same
  class of issue with a different TLS mechanism.

##  Expected behavior

  Either produce correct results (by also disabling inference_mode on
  transform entry, matching the existing no_grad handling) or raise
  an error.

cc @ezyang @gchanan @kadeng @msaroufim @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @Chillee @samdow @kshitij12345

### PR DESCRIPTION
**Pull request by stmcgovern**
Fixes #177318

inference_mode excludes autograd dispatch keys from TLS, causing functorch TensorWrappers to lack autograd metadata. The transforms already handle no_grad by saving prev_grad_mode and enabling grad; this extends the same treatment to inference_mode for grad, vjp,and jvp. Vmap and functionalize are not addressed in this PR. 

Add _disable_inference_mode() to surgically disable inference_mode without clobbering grad_mode/fw_grad_mode (which inference_mode(False) would do, breaking the prev_grad_mode invariant). Wrap grad_increment_nesting, jvp_increment_nesting, and vjp's backward closure. 

My hope is that this can unblock inference_mode workflows where one needs to compute gradients of frozen weights, with respect to inputs(e.g., particle positions) and not model parameters. This happens, for example, in molecular dynamics with a neural net potential function; compute the forces for the particle update steps.

### CODE REVIEW

**Review by aorenste — APPROVED** [review]
Should there be (or is there) a tracking issue for vmap and functionalize?
